cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -362,3 +384,144 @@ func (m *McmManager) GetMachineDeploymentNodes(machinedeployment *MachineDeploym
+
+	md, err := m.machineclient.MachineDeployments(m.namespace).Get(machinedeployment.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("Unable to fetch MachineDeployment object %s, Error: %v", machinedeployment.Name, err)
+	}
+
+	var region string
+	var tags map[string]string
+	var instance instanceType
+	machineClass := md.Spec.Template.Spec.Class
+	switch machineClass.Kind {
+	case "AWSMachineClass":
+		mc, err := m.machineclient.AWSMachineClasses(m.namespace).Get(machineClass.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("Unable to fetch AWSMachineClass object %s, Error: %v", machinedeployment.Name, err)
+		}
+		awsInstance := aws.InstanceTypes[mc.Spec.MachineType]
+		instance = instanceType{
+			InstanceType: awsInstance.InstanceType,
+			VCPU:         awsInstance.VCPU,
+			MemoryMb:     awsInstance.MemoryMb,
+			GPU:          awsInstance.GPU,
+		}
+		region = mc.Spec.Region
+		tags = mc.Spec.Tags
+	case "AzureMachineClass":
+		mc, err := m.machineclient.AzureMachineClasses(m.namespace).Get(machineClass.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("Unable to fetch AzureMachineClass object %s, Error: %v", machinedeployment.Name, err)
+		}
+		azureInstance := azure.InstanceTypes[mc.Spec.Properties.HardwareProfile.VMSize]
+		instance = instanceType{
+			InstanceType: azureInstance.InstanceType,
+			VCPU:         azureInstance.VCPU,
+			MemoryMb:     azureInstance.MemoryMb,
+			GPU:          azureInstance.GPU,
+		}
+		region = mc.Spec.Location
+		tags = mc.Spec.Tags
+	default:
+		return nil, cloudprovider.ErrNotImplemented
+	}
+
+	nodeTmpl := &nodeTemplate{
+		InstanceType: &instance,
+		Region:       region,
+		Zone:         "undefined",
+		Tags:         tags,
+	}
+
+	return nodeTmpl, nil
+}
+
+func (m *McmManager) buildNodeFromTemplate(name string, template *nodeTemplate) (*apiv1.Node, error) {
+	node := apiv1.Node{}
+	nodeName := fmt.Sprintf("%s-%d", name, rand.Int63())
+
+	node.ObjectMeta = metav1.ObjectMeta{
+		Name:     nodeName,
+		SelfLink: fmt.Sprintf("/api/v1/nodes/%s", nodeName),
+		Labels:   map[string]string{},
+	}
+
+	node.Status = apiv1.NodeStatus{
+		Capacity: apiv1.ResourceList{},
+	}
+
+	// Numbers pods per node will depends on the CNI used and the maxPods kubelet config, default is often 100
+	node.Status.Capacity[apiv1.ResourcePods] = *resource.NewQuantity(100, resource.DecimalSI)
+	node.Status.Capacity[apiv1.ResourceCPU] = *resource.NewQuantity(template.InstanceType.VCPU, resource.DecimalSI)
+	node.Status.Capacity[gpu.ResourceNvidiaGPU] = *resource.NewQuantity(template.InstanceType.GPU, resource.DecimalSI)
+	node.Status.Capacity[apiv1.ResourceMemory] = *resource.NewQuantity(template.InstanceType.MemoryMb*1024*1024, resource.DecimalSI)
+
+	node.Status.Allocatable = node.Status.Capacity
+
+	// NodeLabels
+	node.Labels = cloudprovider.JoinStringMaps(node.Labels, extractLabelsFromMachineDeployment(template.Tags))
```

Comment:
Can we please make use of [`NodeTemplateSpec`](https://github.com/gardener/machine-controller-manager/blob/master/pkg/apis/machine/types.go#L78) from MachineDeployment to extract both Labels and Taints.
- Basically, MCM provides a mechanism to set/maintain the labels and taints on the node-objects via MachineDeploymentSpec. We can extract the future labels and taints for the nodes from there. 
- This will allow the node-affinity to work against generic labels and taints on nodes. 
Wdyt?